packages/contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { SignedMath } from "@openzeppelin/contracts/utils/math/SignedMath.sol";
+import { FixedPointMathLib } from "@rari-capital/solmate/src/utils/FixedPointMathLib.sol";
+import { Burn } from "../libraries/Burn.sol";
+
+/**
+ * @title ResourceMetering
+ * @notice ResourceMetering implements an EIP-1559 style resource metering system where pricing
+ * updates automatically based on current demand.
+ */
+contract ResourceMetering {
+    /**
+     * Struct representing current resource parameters.
+     */
+    struct ResourceParams {
+        uint128 prevBaseFee;
+        uint64 prevBoughtGas;
+        uint64 prevBlockNum;
+    }
+
+    /**
+     * Along with the resource limit, determines the target resource limit.
+     */
+    int256 public constant ELASTICITY_MULTIPLIER = 4;
+
+    /**
+     * Denominator that determines max change on fee per block.
+     */
+    int256 public constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    /**
+     * Maximum amount of deposit gas that can be used within this block.
+     */
+    int256 public constant MAX_RESOURCE_LIMIT = 8_000_000;
+
+    /**
+     * Target amount of deposit gas that should be used within this block.
+     */
+    int256 public constant TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER;
+
+    /**
+     * Minimum base fee value, cannot go lower than this.
+     */
+    int256 public constant MINIMUM_BASE_FEE = 10_000;
+
+    /**
+     * Initial base fee value.
+     */
+    uint128 public constant INITIAL_BASE_FEE = 1_000_000_000;
+
+    /**
+     * EIP-1559 style gas parameters.
+     */
+    ResourceParams public params;
+
+    /**
+     * Sets the initial resource values.
+     */
+    constructor() {
+        params = ResourceParams({
+            prevBaseFee: INITIAL_BASE_FEE,
+            prevBoughtGas: 0,
+            prevBlockNum: uint64(block.number)
+        });
+    }
+
+    /**
+     * Meters access to a function based an amount of a requested resource.
+     *
+     * @param _amount Amount of the resource requested.
+     */
+    modifier metered(uint64 _amount) {
+        // Record initial gas amount so we can refund for it later.
+        uint256 initialGas = gasleft();
+
+        // Run the underlying function.
+        _;
+
+        // Update block number and base fee if necessary.
+        uint256 blockDiff = block.number - params.prevBlockNum;
+        if (blockDiff > 0) {
+            // Handle updating EIP-1559 style gas parameters. We use EIP-1559 to restrict the rate
+            // at which deposits can be created and therefore limit the potential for deposits to
+            // spam the L2 system. Fee scheme is very similar to EIP-1559 with minor changes.
+            int256 gasUsedDelta = int256(uint256(params.prevBoughtGas)) - TARGET_RESOURCE_LIMIT;
```

Comment:
I'm worried about this underflowing here

Reply:
Yes, that's the desired behavior, it will go negative